File: src/linked-list/linked-list-reversal.ts
import { ListNodeOrNull } from '../utils/types'

export function linkedListReversal<T>(
    head: ListNodeOrNull<T>
): ListNodeOrNull<T> {
    if (head == null || head.next == null) {
        return head
    }

    let prevNode: ListNodeOrNull<T> = null,
        currentNode: ListNodeOrNull<T> = head
    while (currentNode != null) {
        const nextNode: ListNodeOrNull<T> = currentNode.next

        currentNode.next = prevNode

        prevNode = currentNode
        currentNode = nextNode
    }

    return prevNode
}

File: src/linked-list/linked_list_reversal.go
package linkedlist

import (
	utils "algorithm-puzzle-exercise/src/utils"
)

func LinkedListReversal[T any](head *utils.ListNode[T]) *utils.ListNode[T] {
	if head == nil || head.Next == nil {
		return head
	}

	var prevNode *utils.ListNode[T] = nil
	currentNode := head

	for currentNode != nil {
		nextNode := currentNode.Next

		currentNode.Next = prevNode

		prevNode = currentNode
		currentNode = nextNode
	}

	return prevNode
}
